src/System.Private.Interpreter/src/Internal/Runtime/Interpreter/ILImporter.Interpreter.cs

```diff
@@ -143,57 +153,97 @@ private void ImportCalli(int token)
 
         private void ImportLoadNull()
         {
-            _interpreter.EvaluationStack.Push(new ObjectRefStackItem(null));
+            _interpreter.EvaluationStack.Push(StackItem.FromObjectRef(null));
         }
 
         private void ImportReturn()
         {
-            bool hasReturnValue = _interpreter.EvaluationStack.TryPop(out StackItem stackItem);
-            if (!hasReturnValue)
+            var returnType = _method.Signature.ReturnType;
+            if (returnType.IsVoid)
                 return;
 
-            switch (stackItem.Kind)
+            int integer = 0;
+            StackItem stackItem = PopWithValidation();
+
+            if (stackItem.Kind == StackValueKind.Int32)
+                integer = stackItem.AsInt32();
+
+            TypeFlags category = returnType.Category;
+            switch (category)
             {
-                case StackValueKind.Int32:
-                    _interpreter.SetReturnValue(((Int32StackItem)stackItem).Value);
+                case TypeFlags.Boolean:
+                    _interpreter.SetReturnValue(integer != 0);
+                    break;
+                case TypeFlags.Char:
+                    Debug.Assert(integer >= char.MinValue && integer <= char.MaxValue);
```

Comment:
I would also delete the other asserts.

```
.assembly stack { }

.method static char GetChar()
{
    ldc.i4 65537
    ret
}

.method static int32 Main()
{
    .entrypoint
    call char GetChar()
    ret
}
```

This returns 1 because the value simply gets truncated.

Reply:
Including asserts for `byte`, `sbyte` etc.?

Reply:
Yup, see ECMA-335 spec, section I.12.1.2 Handling of short integer data types. The evaluation stack only deals with 4 or 8 byte values. Values are widened when loaded and truncated when stored.